### ISSUE TITLE
Map this Photo (w direction & sharing capabilities)

### ISSUE THREAD

**Issue by theantichris** [theantichris/BouncyBall#1]
# Map this Photo (w direction & sharing capabilities)

<aside>
🏆 Eliminate guess work for contractors by making “map this photo” highly discoverable, intuitive, and sharable.

</aside>

# 🌊 PROBLEM

---

- CoCam users often report the need for a way to visually show where a photo was taken on a map. (e.x.1, this photo shows an issue that needs to be addressed, but I don’t know *where* on the job site it was taken)
- Today they figure this out by training their team to take photos close up and with a larger context view, but it doesn’t always happen. So the user trying to understand views the images before and after the image of the problem to try and piece together enough context for where the problem exists.
- Here is our research documentation: https://companycam.dovetail.com/insights/dZBUXC82Hg1atmpUp0R1R

# 🎯 DEFINITION OF DONE

---

- Done = code complete and shipped (possibly behind flag) by end of cycle.
    - ✅ First 2-week goal  —  make it discoverable, baseline eventing, understand the variance in accuracy - add to 3/4 RC (no flag)
    - Second 2-week goal  —  open map inside the app, and introduce direction-facing detail
    - Third 2-week goal — ux enhancements, nice-to-have’s, and final testing final efforts will go with 4/1 RC
- Order priorities for how to build S I M P L E
    1. Simple to validate - will this help us expand up market and close more deals? 
    2. Simple for user to use - if it’s too clunky, users won’t find it valuable
    3. Simple to extend - if this is successful, then we’ll likely add more to the map
    4. Simple to build - it doesn’t need to be simple to build, but what we build should be simple

# 📈 SUCCESS

---

- Adoption of the feature - we want to see 20% of companies engaging with this feature more than once a month within the first 3 months of release
- Sales people using this feature to sell (will see increase in “map this photo” reference in Gong conversations)
- Hypothesis: increase in conversation rate for residential and commercial roofing, solar, telecom/util, HVAC, Plumbing, Electric (see [[2023 baseline](https://companycam.metabaseapp.com/question/5466-conversion-rate-by-trade-sum-of-trials-ended-grouped-by-trade-2023)](https://companycam.metabaseapp.com/question/5466-conversion-rate-by-trade-sum-of-trials-ended-grouped-by-trade-2023) here) → were users in who mapped a photo more likely to convert?

# 🤩 SOLUTION

---

## First ~~Two~~ Week~~s~~

1. **~~Move the feature to Premium tier.~~**
    1. ~~Hide it for Pro and Basic user.~~
    - ❌~~Make sure this feature is on the premium tier +++ only~~
2. ✅ **Make the feature discoverable** so that we don't have to announce it, users should understand from looking at a photo that they can plot it on a map. (We can debate whether it actually needs to be this prominent or not)
    1. Button on bottom
        1. tapping button opens native map with pin where photo is
    - ✅Remove original flow completely → @Jordan Small will communicate to S3 that the flow will change with the 3/4 RC
3. ✅ **Disable map button if asset is not mappable (no location data or offline)**
    1. button is disabled
4. ✅ **Hold ~~confidence in the predictable~~ awareness of the variance in accuracy of the “Map this photo” feature** 
    
    Note: this is intended to inform how we might design the feature in a way that a user can trust/feel comfortable with 
    
    - ✅Adjust GPS accuracy / polling distance as necessary
        - May not actually have much control over the accuracy. There appears to be some other data points that go into calculating overall GPS accuracy.
        - We could display the device accuracy (much like how we display stamps for example) when someone takes photos. We have an accuracy number that appears to be meters on iOS and Samsung is much different, possibly feet we’re not sure yet. It’s also possible that Samsung is just way worse on this specific device
    - ✅**predictable accuracy**
        1. let’s reduce the threshold for gps detection from 2m to 1m 
        2. 🍰fix bug where we strip EXIF data from photos on some Android devices
            1.  https://github.com/CompanyCam/companycam-mobile/issues/4714

## Remaining Time

1. **Put the map inside our app** in order to alleviate the confusing experience of finding yourself outside of the app (and to increase our control of the UX).
    - **UX**
        1. show the map within the app
            1. existing pattern with mapping projects on the home screen (map button above the camera button)
            - ✅Migrate the current solution to use an in-app map, instead of the native map.
        2. ~~display the area where the photo might have been taken (margin of error in order to build users’ confidence in the accuracy of the tool)~~
    - Update `react-native-maps` to current version, if applicable.
2. **Get and display the direction the user was facing** in order to help users understand what the user was facing when they took the picture. 
    - **direction the user was facing**
        1. save direction to photo information
        2. 🍰stamp the direction?
        3. 🍰altitude / elevation
        - Modify existing API Endpoint to capture new location information to the DB
            - **Direction** the user was facing (needed)
            - **Altitude** (needed)
            - **~~distanceFilter~~**  (we’ll add this to the log instead)
            - **Accuracy (GPS)**
            - **Altitude Accuracy**
            - ~~Include the previous location / coordinate info - could use this to calculate the direction they’re moving if the “direction facing” is not accurate~~
            - ~~accelerometer~~
            - ~~gyro~~
            - ~~barometer~~
    - Add above data to EXIF data for the image, as necessary.
    - [ ]  dev task from shaping: look at the format of the direction data and decide ~~who will~~ how to convert to the format of the UI  (this needs to be done before scoping on Monday)
    - [ ]  Question: how many roofers are pointing the camera down?  do we detect that? how do we display that?
        - Probably a lot - it’s a use case to think about how to display the direction information in this context
        - It probably wont ever be directly down, like exactly 180 degrees.  So we may need to agree on thresholds
        - accuracy of “direction faced” depends on the accuracy of the gyroscope in the user’s device
        - if we can think of something valuable to do here - we can add it as a 🍰 item; otherwise just take the direction facing information as it is.
3. **Make the UX feel good to use.** 
    1. Increase a user’s tolerance for inaccuracy
        1. Can we reduce confusion/negative emotions for a user when they see a pin is in the wrong spot? @Jason McKelvey 
        2. we will do this with the line of copy on the information modal
    2. Make sure the pin shows up on the map → @Andy Albrecht let’s identify bugs related to this and prioritize them in-cycle
    3. When the photo has no location data, give the user some information:
        1. UX: user taps visually-disabled button:
            1. does what?  shows a tool-tip style message or a toast or something
            2. says what?  This image cannot be mapped. ← @Sarah Akinyemi final copy is due by Friday
            - [ ]  dev task: get translation  (let’s hold on this until Jason says it’s final)
        2. event only fires when we have the location data and actually open the map
    4. When the user is offline:
        1. the button is legitimately disabled (no tool tip) like the other buttons in that row
4. Leverage in-app marketing conventions to **introduce the feature** to users 
    1. show a simple tool tip on the map button 
    2. if never tapped, how long should it appear? 5 seconds
    3. if never tapped, how many times will the tool tip appear? 2x —> if this overcomplicates it, then just show once ever
    4. if tapped once, does it ever appear again? No
    5. should it persist between installs? no (unless doing that is real complicated, in which case, we’ll just remove the code after a month or something)
5. **🍰? Make it (photo+map) sharable** so that as a user I can add it to a report, add it to Pages, or share the single asset (photo + map) with my sub-contractor who isn’t using CompanyCam.
    1. MVP: can we track screenshot taken?
6. 🍰**Get the elevation** and start to explore how we might make that data meaningful/valuable to customers
7. 🍰 **Research spike on visually displaying the direction on the map** (rather than just in text)
8. 🍰 Datadog dashboard that we can use to visualize impact of 

Other related to-do’s but not for scoping:

- [ ]  @Sarah Akinyemi to determine how we’ll measure success with the selling

# 🛠️ Technical Solution

---

- INFORMATION GATHERING
    - Backend questions
        - How are the Images currently stored
            - Stored in Postgres, but `Assets` are stored in Mongo
        - How do we want to handle querying images, as they are currently paginated
            - This will most-likely need to be a new endpoint since pagination won’t work for this feature
        - How do we handle images with no geo info
            - Don’t include them? Guessing we’ll just need to filter these out in the query
        - How should we handle map clustering for large image sets
            - We’ll need to determine the cluster size based on the zoom level
        - How will all this affect API performance? (Fast /  small responses?)
            - Dynamically controlling the cluster size should help, but we can also have sub-query to know the media count (photos & videos)
    - Front End questions
        - Does the map feature that we use now have direction built in if we add it. Do we need to switch map tools?
    - How should we future proof extending the feature set
        - EX: how should we handle videos on the map, etc.
        - How should we think about altitude (elevation data)
            - {lat, lon, elev, thumbnail_url, asset_url (m3u8 / hi resolution photo url)}
        - Distance between photos (or map TO photo)
        - Handling 3D stuff
    - How will this work in OFFLINE mode
    - Costs?
    - @Anonymous have you done anything with heat mapping / clustering of map markers in your current cycle?
    - Potential example return object from BE
        
        ```jsx
        [{
        	lat,
        	lon,
        	thumbnail_url,
        	asset_url,
        	asset_type
        }]
        ```
        

# 🐰 RABBIT HOLES

---

## Questions:

1. What areas of work specified in the solution above could benefit from a BE dev?
    1. Add lat/lon to asset feed item (kindof already in progress)
    2. Determine how to consume altitude / elevation
    3. Determine how to consume photo direction
    4. Create a `jsonb` field for positional information (Type it for validation)
        1. If we go this route, should we index it for querying?
2. Does the accuracy of the gps vary? 
    1. Context: Some folks showed me a pin that was dropped on the roof, but the photo was taken in the back yard (+ other examples). 
    2. If so, what factors make it vary? 
        1. Signal strength
        2. How does power / heat affect GPS?
        3. GPS can be turned off. 
        4. in addition to accuracy, make sure we are looking at the time stamp as well.
            1. EX: if you had gps on, then turned it off, and started taking pics, are we using “cached” or old data?
            2. How tightly should this be integrated with offline stuff?
    3. Can we have consistent confidence interval / margin of error? 
        1. We ping for GPS when a user has moved 2m … should we update to 1m instead in order to achieve approximately a 3 foot margin of error? 
        2. Potentially add a timestamp check, to see if we’re using old data.
    4. Are there times when we just know it’s not going to be accurate? What are those? 
        1. if gps is off
        2. if signal strength is low
            1. Potentially show a warning instead of the map in low accuracy times
    5. ~~specifically what is the nuance between:~~
        1. offline (captured in the app)
            1. **Directly poll the gps for location**
        2. captured externally with location services on
            1. **Relying on exif data which could be inaccurate as it’s mutable.**
                1. What information do we have/not have from internally captured vs. externally captured photos?
                    1. UX question: do we want to visually distinguish photos taken in the app from those taken externally and be able to treat them differently (I.e., we might display one and not the other, or display one as blue and one as green pins, etc)
    6. Do we need to figure out ground level vs. what we get from `altitude` in the gps?
    
3. ~~We would like to using the CoCam internal map (we currently use Google, right?)… what would that take? How much would it take to convert to internal map vs. using native map as we do now?~~
    1. ~~One consideration: do we need to leverage the internal map to facilitate sharing?~~
        1. **YES**
4. ~~Related to detecting the direction a user was facing: can we show that on the native map or would this require us to use something else (like google street view)?~~
    1. **We have options to display this on the map that is internal to CoCam.**
5. ~~Can we show direction-facing from externally uploaded photos~~
    1. I**t depends on whether the camera used is capable of gathering direction** —> how will we handle communicating why this information is missing?
        1. Exif.GPSInfo.GPSImgDirection
        2. [https://opencamera.org.uk](https://opencamera.org.uk/) ← open source resource that Christopher found
        3. https://www.exiftool.org/geotag.html <- exiftool is a good resource from Dan
    2. ~~do externally taken/uploaded photos include direction with EXIF data?~~ 
        1. If the camera the photo was taken with supports direction.
            1. note: relevant with that bug on android which sometimes strips it out
6. What nuances for Android?
    1. Low power mode: GPS accuracy is diminished
7. What nuances for iOS?
    1. Low power mode: GPS accuracy is diminished
    2. Per Christopher: 
        
        Android has a High Accuracy ModeTurn on high accuracy modeTo help Google Maps find your location with the most accurate blue dot, use high accuracy mode.
        
        1. On your Android phone or tablet, open the Settings app .
        2. Tap Location.
        3. At the top, switch location on.
        4. Tap Mode High accuracy.
8. ~~What eventing do we need? → in fact if anyone can put an element clicked event on “map this photo” during cool down, that’d be so rad.~~
    1. ~~Devin is handling this during cool-down.~~
    2. Should also log the location information (altitude, accuracy, heading / direction) when we log upon capturing a photo. I think we should pregame this for videos as well.
9. ~~Exploratory question: Is it possible to take two photos at once (one with wide angle lens and one close up? and composite those?)~~ 
    1. **YES** Andy was able to take 4 photos at once.
10. ~~We’d like to stop opening the external map and start keeping the whole experience in the app. When the user maps the photo, how much map do we want to show?~~
    1. **We can adjust this.**
11. 🍰 What’s the UX for a user who is offline and wants to map a photo? 
    1. current experience is that when a user taps a thumbnail, we show a “looks like this image is missing” → should we display a different message?
    2. ~~when we cache projects for offline: what all do we cache?~~
    3. ~~can we say “for projects that are cached, include the most mapped photos (by all users) in the cache”?~~
12. ~~🍰 Can we gather and display elevation too?~~ 
    1. **Yes but the camera would have to include that information when the photo was taken.**
13. UX questions
    1. ~~would we want to augment the lat/long/direction + accuracy to the db? rather than getting it from EXIF (could pass directly to server when we create the photo)~~
        1. We will save this info in both the EXIF data and the database.
    2. When a user does **not have a native map installed**, they will not be able to view the image plotted on a map. Some users may experience that as we proceed with v1 of “Map this photo” but it will be remedied for them once we’re opening the map in-app.

# 🚫 NO-GOs

---

1. **Mobile only**, we will not be adding this feature to web.
2. **1 photo only**, we will not map multiple photos at a time, nor video
3. **Will not map these special types of photo:**
    - photos without the needed location information
        - including: Before&After, <what else?>
4. We will not prompt users to **enable high-accuracy GPS** for specific images at this time
5. We will not show the map in edit mode. Instead, we will follow current behavior while in edit mode (do not show the map).

# ✅ Shaping Checklist

---

<aside>
💡 Use this as a tool to help you think through the technical considerations of the pitch. Ask both **“How can we make this work?”** and **“How could this fail?”**. Consider how each applies (or doesn’t), identify what investigations or actions are needed, and capture conclusions/decision. There may be additional areas that deserve more thought for this pitch, add them to the list.

</aside>

**Context**

- [x]  Existing Features - What else will this touch within our application? Some things may include Notifications, Collaboration, User Permission Roles, etc.
    - Using the existing map
- [x]  Existing Bugs - What [known bugs](https://github.com/orgs/CompanyCam/projects/29/views/1) are in this area of the code? What complications might they cause? Should we fix any of them before or as part of this work?
    - https://github.com/CompanyCam/companycam-mobile/issues/4714
- [x]  Tech Debt - Identify outdated tech and/or deprecated patterns we are using for this solution and if it makes sense to pay down that debt as a part of this solution. Will we need to refactor anything?
    - We are doing small refactors
    - Backend reference - [Patterns and Libraries](https://www.notion.so/Patterns-and-Libraries-7a4458e6b2e04be8a3a8dad4c681b374?pvs=21)
    - Client reference - https://github.com/CompanyCam/client-patterns-and-practices

**Implementation**

- [x]  Web &/or Mobile - is parity needed or no?
- [x]  Libraries - Will we consider introducing any new libraries as part of this project? (If so, you’ll want a Discipline Lead Review).
    - 🍰 updating the react native map
    - [x]  When considering a new library or build vs. buy decisions, complete a decision matrix comparing solutions. See [Build vs Buy](https://www.notion.so/Build-vs-Buy-12f1a59d19ac469f899fd2a116d9b896?pvs=21) for guidelines.
- [x]  Performance - Under what scenarios could performance be negatively impacted? How common do we expect those scenarios to be? How might we account for that in our technical design? What will we consider “good enough” for performance?
- [x]  Scalability - What limitations or issues might we hit in scaling? e.g. Resource usage, database query speed? How might we account for that in our technical design?
    - @Steven McManus and team are having conversations w @Jeff McFadden  rn about the data we are looking to gather and the impact of it
        - The data in question are [listed above in this pitch](https://www.notion.so/Map-this-Photo-w-direction-sharing-capabilities-bce807a20b714770b7d2fbccb59c1c45?pvs=21).
        - We’re discussing returning these data points:
            - every time a photo is queried for
            - only when a photo is mapped
        - How do we determine whether to show the button as enable or visually-disabled with tooltip?
            - we get that with
    - the API calls to Google Maps API do cost us money, which will scale with increased usage —
        - do we use our API key for these calls?
        - do they cost money?
        - do we have concerns?
- [x]  Sustainability - How might we build on this work in the future? How will we maintain it? Are we making any intentional tradeoffs now? Keep [YAGNI](https://martinfowler.com/bliki/Yagni.html) in mind.
- [ ]  Design System - What existing Design System components and/or variables can we use?
    - using design convension for introducing the button; and giving feedback when the button is disabled
- [ ]  Translations - What translations are needed?
    - copy for “no location data” feedback
    - copy for the introductory tool tip
    - copy for the map view where we say “stuff is inaccurate sometimes”
- [x]  Steven (Eventing) - Has the team reviewed the planned events in the [Snowplow Event Lexicon](https://docs.google.com/spreadsheets/d/1gds5wNokNVoNiMVivfgwIPM06DvJNWEeD8wTnxFieAU/edit#gid=0) and/or [](https://www.notion.so/65c4890255104e6698bda1df62fff9d4?pvs=21) ?
- [x]  QA - How will we test this work?

**Security**

- [x]  How can this improvement be misused?
- [x]  Does this improvement cause unnecessary data exposure?
- [x]  Does this improvement expose an attack vector?
- [x]  Are the data sources used by this improvement trusted or untrusted?
- [x]  Does this improvement follow the privacy principles of:
- [x]  Data minimization? (Is only strictly necessary personal information collected, used, stored, and transferred by this improvement?)
- [x]  Purpose specification? (Is the existence of this improvement, its usage of the personal information and the reasons why it processes the personal information captured by the Privacy Policy?)
- [x]  Use, retention and disclosure? (Does this improvement use the personal information, store it and disclose it in ways which are compatible with the purpose(s) for which it is designed?)

**Reviews**

- [x]  Money - Does this project fall into the [Money Bucket](https://www.notion.so/Money-Bucket-b16f5ac5126744a2ba7a59e4fc983c82?pvs=21)?
    - [x]  If so, have you followed the appropriate protocol?
- [x]  Build vs Buy - Are you considering a solution that involves paying for a 3rd party product/service?
    - [x]  If so, have you followed the process laid out in [Build vs Buy](https://www.notion.so/Build-vs-Buy-12f1a59d19ac469f899fd2a116d9b896?pvs=21)?
- [x]  Discipline Lead Review - Share your proposed solution with your discipline lead if it meets any of the following criteria:
    - Changes to system architecture, including major refactors and greenfield additions.
    - Implementation of a new system level feature.
    - Introducing a new library, dependency.
    - Introduction of new technical pattern.
    - Introduction of a solution that is intended to be used by multiple teams/areas of the codebase (i.e. cross-team solutions).
    - Touches a central, foundational part of the app.
- [ ]  Design System Review - Share your proposed solution with the `#design-system` group if it meets any of the following criteria:
    - [ ]  @Jason McKelvey let @Steven McManus  know when designs are final and can be submitted.
    - Changes to a Slab component
    - Creating/Using components that could be reused outside the feature

**Impacts and Dependencies Outside the Team**

- [x]  Integrations/External API - Might this work impact our existing integrations or external API? If so, pull in the Integrations team.
- [x]  Internal Team Impact - Consider if this will impact current data being utilized by data team and/or sales. Think Metabase, Salesforce, Dash.
- [x]  Communication Plan - Identify teams that might be impacted by this work. Do they need to be consulted or just informed? Make a plan for communicating with them.
    - Sarah will keep Ming-Ming informed since Foundation team is also doing map work.
- [ ]  Documentation - What documentation will we want to include in the scope of this project? e.g. any architecture changes, data models, data flows, etc.
- [x]  Testing - What end-to-end tests need to be updated or added?
- [x]  Infrastructure - Will the work require any changes to our infrastructure or potentially require the support of our Infrastructure team? If so, notify the Infrastructure team and/or submit a request in #infrastructure-requests.

**Production**

- [x]  Backwards Compatibility - How might releasing this feature impact existing users, especially users on older mobile app versions? How will we handle this?
    - Moved where this lives and removed the old flag; we’re accepting the risk to the small number of people
- [x]  Feature Flags - Does this involve work that is behind a feature flag? Do previous flags need to be removed? Is a new flag needed?
- [x]  Observability - How will we monitor performance and usage of this work in production?
    - Sarah will view usage data in Metabase
    - WHO will view performance impact where/how
        - do we need the datadog dashboard for vibe check
- [x]  Internal Tooling - What post-release maintenance and support of the feature will be needed? Do we need to build any new tools for this purpose?
